proof/refine/ARM/Reply_R.thy

```diff
@@ -1081,27 +1024,841 @@ lemma sc_replies_relation_replyNext_update:
+
+lemma update_replyPrev_replyNexts_inv[wp]:
+  "updateReply rp (replyPrev_update prev) \<lbrace>\<lambda>s. P (replyNexts_of s)\<rbrace>"
+  unfolding updateReply_def supply fun_upd_apply[simp del]
+  apply wpsimp
+  by (metis ko_at'_replies_of' map_upd_triv opt_map_upd_Some)
+
+(* replyRemoveTCB_corres specific corres rules *)
+
+lemma setSchedContext_scReply_update_None_corres:
+  "corres dc ((\<lambda>s. (sc_replies_of s |> hd_opt) ptr = Some rp) and valid_objs and pspace_aligned and pspace_distinct)
+             \<top>
+          (update_sched_context ptr (sc_replies_update (takeWhile ((\<noteq>) rp))))
+         (do sc' \<leftarrow> getSchedContext ptr;
+             setSchedContext ptr (scReply_update Map.empty sc')
+          od)"
+  apply (rule_tac Q="sc_at' ptr" in corres_cross_add_guard)
+   apply (fastforce dest!: state_relationD simp: obj_at_def is_sc_obj_def vs_heap_simps
+                    elim!: sc_at_cross valid_objs_valid_sched_context_size)
+  apply (rule corres_symb_exec_r)
+     apply (rule_tac P'="ko_at' sc' ptr" in corres_inst)
+     apply (rule_tac Q="sc_obj_at (objBits sc' - minSchedContextBits) ptr" in corres_cross_add_abs_guard)
+      apply (fastforce dest!: state_relationD ko_at'_cross)
+     apply (rule corres_guard_imp)
+       apply (rule_tac P="(\<lambda>s. (sc_replies_of s |> hd_opt) ptr = Some rp)
+                          and sc_obj_at (objBits sc' - minSchedContextBits) ptr"
+                  and n1="objBits sc' - minSchedContextBits"
+                            in monadic_rewrite_corres[OF _ update_sched_context_rewrite])
+       apply (rule corres_symb_exec_l)
+          apply (rule_tac P="(\<lambda>s. kheap s ptr = Some (kernel_object.SchedContext sc (objBits sc'
+                                                                                    - minSchedContextBits)))
+                              and K (rp = hd (sc_replies sc))"
+                      and P'="ko_at' sc' ptr"  in corres_inst)
+          apply (rule corres_gen_asm')
+          apply (rule corres_guard_imp)
+            apply (rule_tac sc=sc and sc'=sc' in setSchedContext_update_corres; simp?)
+             apply (clarsimp simp: sc_relation_def objBits_simps)+
+         apply (wpsimp wp: get_sched_context_exs_valid simp: is_sc_obj_def obj_at_def)
+          apply (rename_tac ko; case_tac ko; clarsimp)
+         apply simp
+        apply (wpsimp simp: obj_at_def is_sc_obj_def vs_heap_simps)
+       apply (wpsimp wp: get_sched_context_no_fail)
+      apply (clarsimp simp: obj_at_def is_sc_obj_def)
+     apply simp
+    apply wpsimp+
+  done
+
+lemma replyPrevNext_update_commute:
+  "replyPrev_update Map.empty (replyNext_update Map.empty reply)
+   = replyNext_update Map.empty (replyPrev_update Map.empty reply)"
+  by (cases reply; clarsimp)
+
+lemma updateReply_Prev_Next_rewrite:
+  "monadic_rewrite False True (reply_at' rp)
+    (do y <- updateReply rp (replyPrev_update Map.empty);
```

Comment:
idea: it may be possible to strengthen this by replacing each `Map.empty` with `x` and `y`?

Reply:
yes, indeed!